### PR TITLE
Introduce default index with new three index types

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR proposes a default index so that we can now forget about the case when index is missing in Koalas DataFrame  - when Koalas DataFrame is directly created from Spark DataFrame.

There are three types of default index that can be controlled by `DEFAULT_INDEX` environment variable.

- **one-by-one**: It implements an one-by-one sequence by Window function without
    specifying partition. Therefore, it ends up with whole partition in single node.
    This index type should be avoided when the data is large. This is default.
    **TL;DR:** `Window` with `row_number` without partitioning spec
 
- **distributed-one-by-one**: It implements an one-by-one sequence by group-by and
    group-map approach. It still generates a one-by-one sequential index globally.
    If the default index must be an one-by-one sequence in a large dataset, this
    index has to be used.
    **TL;DR:** `groupby(partition_id).count().collect()` and `groupby(partition_id).apply(f)`.

 - **distributed**: It implements a monotonically increasing sequence simply by using
    Spark's `monotonically_increasing_id` function. If the index does not have to be
    a one-by-one sequence, this index should be used. Performance-wise, this index
    almost does not have any penalty comparing to other index types.
    **TL;DR:** `moninically_increasing_id()`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
will fix the tests soon.

File: databricks/koalas/internal.py [hunk 1]
```diff
@@ -359,6 +363,16 @@ def __init__(self, sdf: spark.DataFrame,
         :param column_index_names: Names for each of the index levels.
         """
         assert isinstance(sdf, spark.DataFrame)
+        if index_map is None:
+            # Here is when Koalas DataFrame is created directly from Spark DataFrame.
+            assert column_index is None
+            assert column_index_names is None
+
+            if "__index_level_0__" not in sdf.schema.names:
```
Comment (HyukjinKwon) [inline databricks/koalas/internal.py]: I will make a separate PR to completely disallow no-index Koalas DataFrame. Seems like there are multiple places to fix.

One side question is that how we will handle the roundtrip in Koalas and Spark DataFrame. If Koalas has an index, `to_spark()` loses index information and we have nowhere to store.

I think, until we figure out a way to store index information properly, we should strip index when we convert to Spark DataFrame.

Currently it's kind of funny:

```
>>> import databricks.koalas as ks
>>> ks.DataFrame(ks.DataFrame({'a': [1,2,3]}).to_spark())
   __index_level_0__  a
0                  0  1
1                  1  2
2                  2  3
```

So basically I would like to propose below until we figure out a clever way to roundtrip.

```
>>> ks.DataFrame({'a': [1,2,3]}).to_spark().show()
+---+
|  a|
+---+
|  1|
|  2|
|  3|
+---+
```

Comment (HyukjinKwon) [reply databricks/koalas/internal.py]: cc @ueshin and @rxin.

File: databricks/koalas/frame.py [hunk 2]
```diff
@@ -4596,11 +4596,11 @@ def merge(self, right: 'DataFrame', how: str = 'inner',
...
-        1  baz        3  baz        7
-        2  foo        1  foo        5
-        3  foo        1  foo        8
-        4  foo        5  foo        5
-        5  foo        5  foo        8
+        5  baz        3  baz        7
+        1  foo        1  foo        5
+        2  foo        1  foo        8
+        3  foo        5  foo        5
+        4  foo        5  foo        8
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: This is actually matched with pandas's result since the output is sorted.

File: databricks/koalas/namespace.py [hunk 3]
```diff
@@ -273,7 +273,7 @@ def read_delta(path: str, version: Optional[str] = None, timestamp: Optional[str
     Examples
     --------
     >>> ks.range(1).to_delta('%s/read_delta/foo' % path)
-    >>> ks.read_delta('%s/read_delta/foo' % path)
+    >>> ks.read_delta('%s/read_delta/foo' % path)  # doctest: +SKIP
```
Comment (HyukjinKwon) [inline databricks/koalas/namespace.py]: Those tests are related with https://github.com/databricks/koalas/pull/639/files#r313350742.

I will make a separate PR to fix it.

File: databricks/koalas/internal.py [hunk 4]
```diff
@@ -394,6 +408,80 @@ def __init__(self, sdf: spark.DataFrame,
...
+
+        """
+        default_index_type = os.environ.get("DEFAULT_INDEX", "one-by-one")
+        if default_index_type == "one-by-one":
+            sequential_index = F.row_number().over(
+                Window.orderBy(F.monotonically_increasing_id().asc())) - 1
+            scols = [scol_for(sdf, column) for column in sdf.columns]
+            return sdf.select(sequential_index.alias("__index_level_0__"), *scols)
+        elif default_index_type == "distributed-one-by-one":
+            # 1. Calculates counts per each partition ID. `counts` here is, for instance,
```
Comment (HyukjinKwon) [inline databricks/koalas/internal.py]: Here, actually it mimics `zipWithIndex` in RDD API.

File: databricks/koalas/internal.py [hunk 5]
```diff
@@ -394,6 +408,80 @@ def __init__(self, sdf: spark.DataFrame,
...
+            almost does not have any penalty comparing to other index types.
+
+        """
+        default_index_type = os.environ.get("DEFAULT_INDEX", "one-by-one")
+        if default_index_type == "one-by-one":
+            sequential_index = F.row_number().over(
+                Window.orderBy(F.monotonically_increasing_id().asc())) - 1
+            scols = [scol_for(sdf, column) for column in sdf.columns]
+            return sdf.select(sequential_index.alias("__index_level_0__"), *scols)
+        elif default_index_type == "distributed-one-by-one":
```
Comment (HyukjinKwon) [inline databricks/koalas/internal.py]: Here, actually it mimics `zipWithIndex` in RDD API.

File: databricks/koalas/internal.py [hunk 6]
```diff
@@ -394,6 +408,80 @@ def __init__(self, sdf: spark.DataFrame,
...
+        environment variable.
+
+        - one-by-one: It implements an one-by-one sequence by Window function without
+            specifying partition. Therefore, it ends up with whole partition in single node.
+            This index type should be avoided when the data is large. This is default.
+
+        - distributed-one-by-one: It implements an one-by-one sequence by group-by and
+            group-map approach. It still generates a one-by-one sequential index globally.
+            If the default index must be an one-by-one sequence in a large dataset, this
+            index has to be used.
```
Comment (ueshin) [inline databricks/koalas/internal.py]: Maybe we should notice that, if we add more data to the data source after creating this index, then it doesn't guarantee the sequential index?

File: databricks/koalas/internal.py [hunk 7]
```diff
@@ -394,6 +408,80 @@ def __init__(self, sdf: spark.DataFrame,
...
+
+        - distributed: It implements a monotonically increasing sequence simply by using
+            Spark's `monotonically_increasing_id` function. If the index does not have to be
+            a one-by-one sequence, this index should be used. Performance-wise, this index
+            almost does not have any penalty comparing to other index types.
```
Comment (ueshin) [inline databricks/koalas/internal.py]: Also notice that we can't use this type of index for combining two dataframes?

File: databricks/koalas/internal.py (hunk 1 above)
Comment (HyukjinKwon) [reply databricks/koalas/internal.py]: But to do that, we should keep index mapping information somewhere. After converting Koalas DataFrame into Spark DataFrame, it's being lost.

Comment (ueshin) [reply databricks/koalas/internal.py]: yeah, loosing the mapping should be okay, I just meant that if the index is named, then the name should be the column name for the Spark DataFrame.

Comment (HyukjinKwon) [reply databricks/koalas/internal.py]: Oh, oh right. Yes.

**Comment by HyukjinKwon** [comment]
I am merging this too. Let me know if you guys have more comments.
